packages/grafana-ui/src/types/datasource.ts

```diff
@@ -4,15 +4,24 @@ import { PluginMeta } from './plugin';
 import { TableData, TimeSeries, SeriesData } from './data';
 import { PanelData } from './panel';
 
-export class DataSourcePlugin<TQuery extends DataQuery = DataQuery> {
+export interface DataSourcePluginOptionsEditorProps<TOptions> {
+  options: TOptions;
+  onOptionsChange: (options: TOptions) => void;
+}
+export class DataSourcePlugin<TOptions, TQuery extends DataQuery = DataQuery> {
   DataSourceClass: DataSourceConstructor<TQuery>;
   components: DataSourcePluginComponents<TQuery>;
+  editor?: React.ComponentType<DataSourcePluginOptionsEditorProps<TOptions>>;
 
   constructor(DataSourceClass: DataSourceConstructor<TQuery>) {
     this.DataSourceClass = DataSourceClass;
     this.components = {};
   }
 
+  setEditor(editor: React.ComponentType<DataSourcePluginOptionsEditorProps<TOptions>>) {
```

Comment:
Editor is a bit too non-descriptive for a data source, as it already as a QueryEditor. 

Think setConfigEditor is more descriptive (as the current angular version is called ConfigCtrl). Also think this should be on the components sub object along with all the other components.